### PR TITLE
Allow reverse tunnel join without exposing the web API

### LINKED ISSUE #12730: Allow agent joining over port 3024
## Problem

Currently, agents joining over reverse tunnel require access to several web handler methods which also serves web UI (port 3080 by default). The methods that they need are /webapi/ping, /webapi/ping, /webapi/host/credentials.

Because the same handler also serves all methods for web UI, there is no way to open up only parts of it, for example, to allow nodes to join the cluster but without also exposing web UI.

## Solution

Agents should be able to join over reverse tunnel port (3024 by default) without requiring access to the full web handler. Reverse tunnel listener should multiplex connections and expose a lightweight version of the web handler with methods mentioned above that agents can use for initial discovery.

With this in place, users will be able to only open 3024 for their agents and not expose entire web UI running on 3080.

### PR DESCRIPTION
**Pull request by atburke**
This PR allows agents to join over a reverse tunnel (port 3024 by default) only, instead of also requiring access to the web API (port 3080). This is accomplished by multiplexing a small subset of the web API (specifically, `/webapi/find` and `/webapi/host/credentials`) over port 3024 when the web API and reverse tunnel aren't already being multiplexed.

Resolves #12730.

### CODE REVIEW

File: lib/web/apiserver.go [hunk 1]
```diff
@@ -283,160 +289,163 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
...
 
-	h.GET("/webapi/github", h.WithAuth(h.getGithubConnectorsHandle))
-	h.PUT("/webapi/github", h.WithAuth(h.upsertGithubConnectorHandle))
-	h.POST("/webapi/github", h.WithAuth(h.upsertGithubConnectorHandle))
-	h.DELETE("/webapi/github/:name", h.WithAuth(h.deleteGithubConnector))
+		// ping endpoint is used to check if the server is up. the /webapi/ping
+		// endpoint returns the default authentication method and configuration that
+		// the server supports. the /webapi/ping/:connector endpoint can be used to
+		// query the authentication configuration for a specific connector.
+		h.GET("/webapi/ping", httplib.MakeHandler(h.ping))
```
Comment (strideynet) [inline lib/web/apiserver.go:305]: I see the original issue states that `/webapi/ping` is needed, but I assume this isn't actually the case.

**Review by strideynet — APPROVED** [review]
Looks good.

File: lib/web/apiserver.go (hunk 1 above)
Comment (atburke) [reply lib/web/apiserver.go:305]: Right. The agent uses `/webapi/find` instead.

File: docs/pages/setup/admin/adding-nodes.mdx [hunk 2]
```diff
@@ -189,12 +189,7 @@ have workloads split across different networks or clouds.
 </Notice>
 
 Using the ports in Teleport's default configuration, the Node needs to be able
-to talk to ports `3080` and `3024` on the Proxy Service. Port `3080` is used to
-initially fetch the credentials (SSH and TLS certificates) and for discovering
-the reverse tunnel. Port `3024 `is used to establish a connection to the Auth
-Service through the Proxy.
-
-For those using ACME, port `443` is also required. 
+to talk to port `3024` on the Proxy Service. For those using ACME, port `443` is also required.
```
Comment (ptgott) [inline docs/pages/setup/admin/adding-nodes.mdx:192]: I think it would help the reader to include some text about what port `3024` is used for, similarly to the original paragraph.

File: lib/service/service.go [hunk 3]
```diff
@@ -3206,6 +3233,53 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
 			log.Info("Exited.")
 			return nil
 		})
+
+		if listeners.minimalReverseTunnelWebEnabled {
```
Comment (codingllama) [inline lib/service/service.go:3237]: Maybe check for `listeners.reverseTunnelMux != nil` and ditch the attribute? It's one less variable to get out of sync.

File: lib/service/service.go [hunk 4]
```diff
@@ -2865,10 +2870,29 @@ func (process *TeleportProcess) setupProxyListeners(networkingConfig types.Clust
...
+				listeners.reverseTunnel, err = process.importOrCreateListener(listenerProxyTunnel, cfg.Proxy.ReverseTunnelListenAddr.Addr)
+				if err != nil {
+					listeners.Close()
+					return nil, trace.Wrap(err)
+				}
+			} else {
+				listeners.minimalReverseTunnelWebEnabled = true
+				listener, err := process.importOrCreateListener(listenerProxyTunnel, cfg.Proxy.ReverseTunnelListenAddr.Addr)
+				if err != nil {
+					return nil, trace.Wrap(err)
```
Comment (codingllama) [inline lib/service/service.go:2883]: Missed listeners.Close()?

```suggestion
					listeners.Close()
					return nil, trace.Wrap(err)
```

File: lib/service/service.go [hunk 5]
```diff
@@ -3206,6 +3233,53 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
 			log.Info("Exited.")
 			return nil
 		})
+
+		if listeners.minimalReverseTunnelWebEnabled {
+			minimalListener, err := multiplexer.NewWebListener(multiplexer.WebListenerConfig{
```
Comment (codingllama) [inline lib/service/service.go:3238]: Could we move the contents of this block to a named func? `initMinimalReverselTunnel()` or something else you prefer?

File: docs/pages/setup/admin/adding-nodes.mdx (hunk 2 above)
Comment (codingllama) [inline docs/pages/setup/admin/adding-nodes.mdx:192]: Is port `3080` a viable alternative to `3024`? Could we use either?

File: lib/web/apiserver.go [hunk 6]
```diff
@@ -531,16 +540,19 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
...
-		AuthClient:    cfg.ProxyClient,
-		AccessPoint:   cfg.AccessPoint,
-		ProxyClient:   cfg.Proxy,
-		CipherSuites:  cfg.CipherSuites,
-		WebPublicAddr: resp.SSH.PublicAddr,
-	})
-	if err != nil {
-		return nil, trace.Wrap(err)
+	var appHandler *app.Handler
+	if !cfg.MinimalReverseTunnelRoutesOnly {
```
Comment (codingllama) [inline lib/web/apiserver.go:544]: nit: I'd rather test for the positive here too (ie, handle minimal tunnels inside the if), but this is minor compared to the one above.

File: lib/web/apiserver.go [hunk 7]
```diff
@@ -283,160 +289,163 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
...
+		// Unauthenticated access to the message of the day
+		h.GET("/webapi/motd", httplib.MakeHandler(h.motd))
 
-	// Desktop access endpoints.
-	h.GET("/webapi/sites/:site/desktops", h.WithClusterAuth(h.clusterDesktopsGet))
-	h.GET("/webapi/sites/:site/desktops/:desktopName", h.WithClusterAuth(h.getDesktopHandle))
-	// GET /webapi/sites/:site/desktops/:desktopName/connect?access_token=<bearer_token>&username=<username>&width=<width>&height=<height>
-	h.GET("/webapi/sites/:site/desktops/:desktopName/connect", h.WithClusterAuth(h.desktopConnectHandle))
-	// GET /webapi/sites/:site/desktopplayback/:sid?access_token=<bearer_token>
-	h.GET("/webapi/sites/:site/desktopplayback/:sid", h.WithAuth(h.desktopPlaybackHandle))
```
Comment (codingllama) [inline lib/web/apiserver.go:439]: Note: I haven't eyeballed every single endpoint, as I think we may refactor this a bit before the end.

**Review by codingllama — COMMENTED** [review]
Looks reasonable to me.

I'm not super familiar with the init functions and the complexity over at some funcs, like setupProxyListeners, is pretty high (not that this is your fault!). Because of that I think we should ask someone like @r0mant to take a look as well.

Apologies for the delay  🙏

File: integration/integration_test.go [hunk 8]
```diff
@@ -6480,3 +6481,38 @@ func testListResourcesAcrossClusters(t *testing.T, suite *integrationTestSuite)
...
+		tconf.Hostname = Host
+		tconf.Token = "token"
+
+		tconf.Auth.Enabled = false
+		tconf.Proxy.Enabled = false
+		tconf.SSH.Enabled = true
+
+		return tconf
+	}
+	_, err := main.startNode(nodeConfig(), main.GetPortReverseTunnel())
```
[the remark this answers is not in the document]
Comment (atburke) [reply integration/integration_test.go:6516]: Yes. Without the other changes, the new node can't fetch what it needs from the web api (host credentials, etc.) from only the reverse tunnel port.

**Review by codingllama — COMMENTED** [review]
LGTM, barring that one small comment.

**Review by codingllama — APPROVED** [review]

**Comment by atburke** [comment]
@ptgott Mind looking at the docs again when you have a moment?

File: docs/pages/setup/admin/adding-nodes.mdx [hunk 9]
```diff
@@ -189,10 +189,10 @@ have workloads split across different networks or clouds.
 </Notice>
 
 Using the ports in Teleport's default configuration, the Node needs to be able
-to talk to ports `3080` and `3024` on the Proxy Service. Port `3080` is used to
-initially fetch the credentials (SSH and TLS certificates) and for discovering
-the reverse tunnel. Port `3024 `is used to establish a connection to the Auth
-Service through the Proxy.
+to talk to port `3024` on the Proxy Service, which is used to both fetch the
```
Comment (ptgott) [inline docs/pages/setup/admin/adding-nodes.mdx:192]:
```suggestion
to talk to port `3024` on the Proxy Service, which is used to both fetch
```

File: docs/pages/setup/admin/adding-nodes.mdx [hunk 10]
```diff
@@ -189,10 +189,10 @@ have workloads split across different networks or clouds.
...
+credentials (SSH and TLS certificates) and establish a connection to the Auth
+Service through the Proxy. Alternatively, port `3080` can be used to fetch
```
Comment (ptgott) [inline docs/pages/setup/admin/adding-nodes.mdx:194]:
```suggestion
Service. Alternatively, port `3080` can be used to fetch
```
"...the Proxy Service, which is used to ... establish a connection to the Auth Service through the Proxy" sounded a bit off to me, since "the Proxy Service" is the subject of this clause.

Comment (ptgott) [inline docs/pages/setup/admin/adding-nodes.mdx:194]: Also, if there's no difference between using port 3024 and 3080, we could probably remove the last sentence of this paragraph and say, "...the Node needs to be able to talk to either port `3024` or `3080` on the Proxy Service...".

Comment (atburke) [reply docs/pages/setup/admin/adding-nodes.mdx]: 3024 is used in both cases to establish the reverse tunnel connection; it's just a question of whether the node fetches credentials over 3024 or 3080. It would have to be something like "either port `3024` or `3024` AND `3080`", and I'm not sure how to word that elegantly.

Comment (ptgott) [reply docs/pages/setup/admin/adding-nodes.mdx]: How does the new behavior work with TLS multiplexing? This paragraph (and the ones after it) seem to assume that TLS multiplexing is disabled by default, while in all supported Teleport versions now (v8 and above), this is not the case.

In any case, I think this paragraph would be much clearer by:

- Replacing specific port numbers with references to Proxy Service configuration settings. Otherwise, the text gets a bit muddled, since we have to go back and forth between port numbers and their usage. If readers need to know the exact ports to use, we could include a command to hit the `/webapi/ping` endpoint.

- Rephrasing "can be used" in the last sentence of the paragraph to something more specific. I'm not sure if 3080 only needs to be _open_ on the Proxy Service or if there's some specific config setting/flag/env variable the user needs to change from something else to 3080.

**Comment by ptgott** [comment]
@atburke If my review is holding up merging the code portion of this PR, please feel free to remove the code change from this PR and open a separate docs PR!

**Comment by codingllama** [comment]
@atburke, I see we're landing new commits here. Should we make another review pass? Should we improve test coverage?

**Comment by atburke** [comment]
@codingllama That seems like a good idea. Once I fix the current test failures I'll re-request.

**Review by codingllama — CHANGES_REQUESTED** [review]
Marking as "request changes" for now. Let us know when you are ready for another pass, @atburke !

**Comment by atburke** [comment]
@codingllama @strideynet I've fixed the last bugs. Feel free to make another review pass when you can.

File: lib/multiplexer/multiplexer.go [hunk 11]
```diff
@@ -379,7 +379,7 @@ func isHTTP(in []byte) bool {
 		[]byte("PATCH"),
 	}
 	for _, verb := range methods {
-		if bytes.HasPrefix(verb, in) {
+		if bytes.HasPrefix(in, verb) {
```
Comment (strideynet) [inline lib/multiplexer/multiplexer.go:382]: Good catch on this bug, I spotted this in my logs recently and never got around to tracking it down.

File: lib/service/service.go [hunk 12]
```diff
@@ -3126,6 +3141,27 @@ func (process *TeleportProcess) setupProxyListeners(networkingConfig types.Clust
...
+		EnableProxyProtocol: cfg.Proxy.EnableProxyProtocol,
+		Listener:            listener,
+		ID:                  teleport.Component(teleport.ComponentProxy, "tunnel", "web", process.id),
+	})
+	if err != nil {
+		listener.Close()
+		return trace.Wrap(err)
+	}
+	listeners.reverseTunnel = listeners.reverseTunnelMux.SSH()
+	go listeners.reverseTunnelMux.Serve()
```
Comment (strideynet) [inline lib/service/service.go:3161]: I know we do a fairly awful job with the existing muxes and this, but it would be nice to log the error potentially returned from `Serve()`.

File: lib/web/apiserver.go [hunk 13]
```diff
@@ -289,16 +293,147 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
...
+			}
+
+			session := struct {
+				Session string
+				XCSRF   string
+			}{
+				XCSRF: csrfToken,
+			}
+
+			ctx, err := h.AuthenticateRequest(w, r, false)
```
Comment (strideynet) [inline lib/web/apiserver.go:359]: Will the error returned from this ever be useful to someone debugging an issue ? Maybe we want to at least log it ?

**Review by strideynet — COMMENTED** [review]
Just a few thoughts, once these are addressed/discussed I'll approve.

**Comment by atburke** [comment]
@codingllama @strideynet Can I get a review on this?

File: lib/service/service.go [hunk 14]
```diff
@@ -3126,6 +3141,31 @@ func (process *TeleportProcess) setupProxyListeners(networkingConfig types.Clust
...
+		ID:                  teleport.Component(teleport.ComponentProxy, "tunnel", "web", process.id),
+	})
+	if err != nil {
+		listener.Close()
+		return trace.Wrap(err)
+	}
+	listeners.reverseTunnel = listeners.reverseTunnelMux.SSH()
+	go func() {
+		if err := listeners.reverseTunnelMux.Serve(); err != nil {
+			process.Config.Log.WithError(err).Debug("Minimal reverse tunnel mux exited with error")
```
Comment (codingllama) [inline lib/service/service.go:3163]: Can we use `process.log` here?

File: lib/multiplexer/multiplexer.go (hunk 11 above)
Comment (codingllama) [reply lib/multiplexer/multiplexer.go:382]: Let's add test coverage for this bug, if we've fallen for it once then it's worth it.

File: integration/integration_test.go [hunk 15]
```diff
@@ -6547,6 +6548,42 @@ func testListResourcesAcrossClusters(t *testing.T, suite *integrationTestSuite)
 	}
 }
 
+func testJoinOverReverseTunnelOnly(t *testing.T, suite *integrationTestSuite) {
+	lib.SetInsecureDevMode(true)
+	defer lib.SetInsecureDevMode(false)
+
+	// Create a Teleport instance with Auth/Proxy.
+	mainConfig := func() *service.Config {
```
Comment (codingllama) [inline integration/integration_test.go:6556]: Why do we need the wrapping func? Could we do `mainConfig := suite.defaultServiceConfig()` and take it from there?

File: integration/integration_test.go [hunk 16]
```diff
@@ -6547,6 +6548,42 @@ func testListResourcesAcrossClusters(t *testing.T, suite *integrationTestSuite)
...
+
+		tconf.SSH.Enabled = false
+
+		return tconf
+	}
+	main := suite.NewTeleportWithConfig(t, nil, nil, mainConfig())
+	t.Cleanup(func() { require.NoError(t, main.StopAll()) })
+
+	// Create a Teleport instance with a Node.
+	nodeConfig := func() *service.Config {
```
Comment (codingllama) [inline integration/integration_test.go:6572]: Same as above.

File: integration/integration_test.go [hunk 17]
```diff
@@ -6547,6 +6548,42 @@ func testListResourcesAcrossClusters(t *testing.T, suite *integrationTestSuite)
...
+		tconf.Token = "token"
+
+		tconf.Auth.Enabled = false
+		tconf.Proxy.Enabled = false
+		tconf.SSH.Enabled = true
+
+		return tconf
+	}
+	_, err := main.StartNodeWithTargetPort(nodeConfig(), helpers.PortStr(t, main.ReverseTunnel))
+	require.NoError(t, err)
```
Comment (codingllama) [inline integration/integration_test.go:6584]: nit: add error details to the focus of the test. It helps draw attention to what is setup vs what is under test, plus it makes an eventual error nicer.

```suggestion
	require.NoError(t, err, "Node failed to join over reverse tunnel")
```

File: lib/reversetunnel/agentpool.go [hunk 18]
```diff
@@ -675,7 +675,8 @@ func (c *agentPoolRuntimeConfig) updateRemote(ctx context.Context, addr *utils.N
 		if ok := errors.As(err, &tls.RecordHeaderError{}); !ok {
 			return trace.Wrap(err)
 		}
-	} else {
+		// Only use the ping results if they weren't from a minimal api handler on the reverse tunnel port.
+	} else if addr.Addr != ping.Proxy.SSH.TunnelListenAddr || ping.Proxy.SSH.TunnelListenAddr == ping.Proxy.SSH.WebListenAddr {
```
Comment (codingllama) [inline lib/reversetunnel/agentpool.go:679]: I'm finding this hard to follow.

1. What is it about the minimal API handler that makes us unable to do TLS routing?
2. Could we get some unambiguous data from Ping that makes this situation easier to detect?

File: lib/service/service.go [hunk 19]
```diff
@@ -3808,6 +3869,62 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
...
+	})
+
+	minimalWebServer = &http.Server{
+		Handler:           minimalProxyLimiter,
+		ReadHeaderTimeout: apidefaults.DefaultDialTimeout,
+		ErrorLog:          utils.NewStdlogger(log.Error, teleport.ComponentReverseTunnelServer),
+	}
+	process.RegisterCriticalFunc("proxy.reversetunnel.web", func() error {
+		utils.Consolef(cfg.Console, log, teleport.ComponentProxy, "Minimal web proxy service %s:%s is starting on %v.",
+			teleport.Version, teleport.Gitref, cfg.Proxy.ReverseTunnelListenAddr.Addr)
```
Comment (codingllama) [inline lib/service/service.go:3916]: nit: break it a bit more for readability?

```suggestion
		utils.Consolef(
			cfg.Console, log, teleport.ComponentProxy,
			"Minimal web proxy service %s:%s is starting on %v.",
			teleport.Version, teleport.Gitref, cfg.Proxy.ReverseTunnelListenAddr.Addr)
```

File: lib/service/service.go [hunk 20]
```diff
@@ -3808,6 +3869,62 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
...
+	minimalWebHandler, err = web.NewHandler(webConfig)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	minimalProxyLimiter.WrapHandle(minimalWebHandler)
+
+	process.RegisterCriticalFunc("proxy.reversetunnel.tls", func() error {
+		log.Infof("TLS multiplexer is starting on %v.", cfg.Proxy.ReverseTunnelListenAddr.Addr)
+		if err := minimalListener.Serve(); !trace.IsConnectionProblem(err) {
+			log.WithError(err).Warn("TLS multiplexer error.")
```
Comment (codingllama) [inline lib/service/service.go:3903]: Shouldn't we return this error instead of swallowing it?

File: lib/service/service.go [hunk 21]
```diff
@@ -3808,6 +3869,62 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
...
+		ReadHeaderTimeout: apidefaults.DefaultDialTimeout,
+		ErrorLog:          utils.NewStdlogger(log.Error, teleport.ComponentReverseTunnelServer),
+	}
+	process.RegisterCriticalFunc("proxy.reversetunnel.web", func() error {
+		utils.Consolef(cfg.Console, log, teleport.ComponentProxy, "Minimal web proxy service %s:%s is starting on %v.",
+			teleport.Version, teleport.Gitref, cfg.Proxy.ReverseTunnelListenAddr.Addr)
+		log.Infof("Minimal web proxy service %s:%s is starting on %v.", teleport.Version, teleport.Gitref, cfg.Proxy.ReverseTunnelListenAddr.Addr)
+		defer minimalWebHandler.Close()
+		if err := minimalWebServer.Serve(minimalListener.Web()); err != nil && err != http.ErrServerClosed {
+			log.Warningf("Error while serving web requests: %v", err)
```
Comment (codingllama) [inline lib/service/service.go:3920]: As above, shouldn't we return this error instead of swallowing it?

File: lib/web/apiserver.go (hunk 6 above)
Comment (codingllama) [reply lib/web/apiserver.go:398]: Ping? Not addressed?

File: lib/web/apiserver.go [hunk 22]
```diff
@@ -289,16 +293,149 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
...
+				if err == nil {
+					out, err := json.Marshal(resp)
+					if err == nil {
+						session.Session = base64.StdEncoding.EncodeToString(out)
+					}
+				} else {
+					h.log.WithError(err).Debug("Could not authenticate.")
+				}
+			} else {
+				h.log.WithError(err).Debug("Could not authenticate.")
```
Comment (codingllama) [inline lib/web/apiserver.go:371]: The nesting in this entire function is making me pull me hair out. Since, I assume, this is mostly moving legacy code I think we can leave as-is, but pulling bits of the code out into separate functions and doing early returns on errors would make wonders here.

**Review by codingllama — CHANGES_REQUESTED** [review]
@atburke, please sync with master and solve the conflicts.

File: lib/reversetunnel/agentpool.go (hunk 18 above)
Comment (atburke) [reply lib/reversetunnel/agentpool.go:679]: The minimal API handler only exists when the proxy and reverse tunnel are listening on separate ports, so it will never do TLS routing.

File: lib/service/service.go (hunk 20 above)
Comment (atburke) [reply lib/service/service.go:3938]: All of the other listeners log the error instead of returning it (and all that returning the error does is log "Teleport process has exited with error." after shutdown). I'd say that we should either log all the errors or return all the errors. Do you have a preference?

File: lib/web/apiserver.go (hunk 6 above)
Comment (atburke) [reply lib/web/apiserver.go:383]: Sorry, I should have said something earlier. I didn't change this one because there's nothing to do in the positive case.

File: lib/service/service.go (hunk 20 above)
Comment (strideynet) [reply lib/service/service.go:3938]: Surely this has more than affect what is output after shutdown? As in, if one of these multiplexers or listeners fails to start and if we return the error, it will actually shut down Teleport, whereas just logging will allow Teleport to continue to run.

**Review by strideynet — APPROVED** [review]
I'm happy to approve this as the majority of this looks good, would be nice to iron out a decision on returning vs warning on errors for the listeners/muxers though. I think that we probably aren't handling this very well in the existing codebase so it might be nice to make a decision and then apply it retrospectively to all the handlers in another PR.

File: lib/service/service.go (hunk 20 above)
Comment (codingllama) [reply lib/service/service.go:3938]: Fair enough, let's not change too much here. We can address it in a separate PR, like Noah suggested below.

File: lib/multiplexer/multiplexer_test.go [hunk 23]
```diff
@@ -868,3 +868,25 @@ func (noopListener) Close() error {
 func (l noopListener) Addr() net.Addr {
 	return l.addr
 }
+
+func TestIsHTTP(t *testing.T) {
```
Comment (codingllama) [inline lib/multiplexer/multiplexer_test.go:872]: Nice, thanks!

File: lib/reversetunnel/agentpool.go (hunk 18 above)
Comment (codingllama) [reply lib/reversetunnel/agentpool.go:679]: Let's simplify and incorporate your response as a comment. See my suggested change below.

File: lib/reversetunnel/agentpool.go [hunk 24]
```diff
@@ -675,7 +675,8 @@ func (c *agentPoolRuntimeConfig) updateRemote(ctx context.Context, addr *utils.N
...
 		tlsRoutingEnabled = ping.Proxy.TLSRoutingEnabled
 	}
```
Comment (codingllama) [inline lib/reversetunnel/agentpool.go:681]: Suggestion: simplify flow / incorporate info about minimal handler.

```suggestion
	}

	// Only use the ping results if they weren't from a minimal handler.
	// The minimal API handler only exists when the proxy and reverse tunnel are
	// listening on separate ports, so it will never do TLS routing.
	isMinimalHandler := /* write condition here */
	if !isMinimalHandler {
		tlsRoutingEnabled = ping.Proxy.TLSRoutingEnabled
	}
```

**Review by codingllama — APPROVED** [review]
Please address the remaining comment, otherwise LGTM.
